RGW: AWS Golang and Java client codes [GSoC' 2022]

pchengma opened the pull request:
This PR includes AWS Golang and Java client codes for my GSoC 2022 application.

Users could follow the steps in README.md files under ceph/examples/java/ and ceph/examples/golang/ to try the example codes.

Signed-off-by: Pengcheng Ma pchengma.enoch@gmail.com

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

yuvalif commented on examples/golang/append_object.go:
```diff
@@ -0,0 +1,89 @@
+
+	// Create a S3 session:
+	s3session := s3.New(session.Must(session.NewSession(&aws.Config{
+		Region:           aws.String("default"),
```
credentials are taken automatically from the aws cli configuration.
you should pass them explicitly here (you can hardcode that in the main functions together with the rest of the vstart specific parameters)

pchengma replied:
Thanks very much for the comments! I will hardcode the credential parameters in the main function.

yuvalif commented on examples/golang/append_object.go:
```diff
@@ -0,0 +1,89 @@
+	// Create a S3 session:
+	s3session := s3.New(session.Must(session.NewSession(&aws.Config{
+		Region:           aws.String("default"),
+		Endpoint:         aws.String("http://127.0.0.1:8000"),
```
this endpoint is specific to vstart. you should pass that to this function (you can hardcode that in the main function)

pchengma replied:
Thanks very much for the comments! I will hardcode the endpoint parameter in the main function. 

I think it is better to hardcode all vstart parameters in the main function and pass s3session as a parameter for uploadObject() and listObject() functions?

yuvalif replied:
sounds good

yuvalif commented on examples/golang/README.md:
```diff
@@ -0,0 +1,41 @@
+
+## Introduction
+
+This directory contains examples on how to use [AWS SDK for Go V1](https://github.com/aws/aws-sdk-go) to experience the RadosGW extensions with the S3 API.
```
these are not extensions. this is the standard S3 functionality

pchengma replied:
Thanks very much for the comments! I will modify the introduction in the README file.

yuvalif commented on examples/golang/README.md:
```diff
@@ -0,0 +1,41 @@
+
+## Feature
+
+We support the following features in our example.
```
you also create the bucket here

pchengma replied:
Thanks very much for the comments! I will add the 'create a bucket' feature in the README file and a createBucket() function in the append_object.go file.

yuvalif commented on examples/java/src/main/java/AppendObject.java:
```diff
@@ -0,0 +1,89 @@
+
+        // Create a S3 client:
+        final AmazonS3 s3 = AmazonS3ClientBuilder.standard().
+                withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, region)).
```
see: https://github.com/ceph/ceph/pull/46083#discussion_r862437764

pchengma replied:
Thanks very much for the comments! I add the hardcoded credential parameter when s3 variable is initialized.

yuvalif commented on examples/java/src/main/java/AppendObject.java:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    // This is a utility function to help create a bucket if not existed.
```
why do you need to check if the bucket already exists before you create it?

yuvalif replied:
also, why do you first look for it manually in the list of buckets, and then use the API to check if it exists?

pchengma replied:
Thanks very much for the comments!

For the first question, I suppose that the user may already create a bucket named mybucket, and if the bucket already exists, we cannot create a bucket with the same name, which will show an error message and stop the following procedure (uploading a bucket). Therefore, I need to check before creating a bucket. It seems there is no such problem in Golang SDK. In Golang SDK, if there exists a bucket with the same name, it will proceed with the following procedures. I need to further check this.

For the second question, I forgot to remove the manually listing part when I learned the AWS Java SDK. I will remove it in the modification.

yuvalif commented on examples/java/src/main/java/AppendObject.java:
```diff
@@ -0,0 +1,89 @@
+                build();
+
+        // Create a bucket if not exists:
+        createBucket(bucket_name, s3);
```
unless you use exceptions you should probably check for error conditions and stop the function if one of the calling functions failed

pchengma replied:
Thanks very much for the comments! I will add exceptions for each function in the modification.

yuvalif replied:
you can use either exceptions or return values (both work in java)
if you use exception don't forget to handle them in the caller

pchengma replied:
Well received! Let me first try both methods, then see the outcomes, and choose the better one between these two methods to keep the example codes easy to understand and elegant for users.

pchengma commented:
@yuvalif Hi Yuval, I have revised the codes according to the comments. Could you please review my codes when you have time? Thanks very much in advance!

Revision Overview:
- Golang
    + Hardcoded credential parameters.
    + Passed s3.S3 as a parameter for uploadObject(), createBucket(), and listObject() functions.
    + Corrected the concept in the Introduction section of README.md file.
    + Added "Create a bucket" to the Feature section of README.md file.
    + Double-checked that we do not need to implement a check whether the bucket with the same name has already existed, since the CreateBucket() function in AWS Golang SDK will continue to upload the new object to the existed bucket without affecting the other objects in that bucket.
- Java
    + Hardcoded credential parameters.
    + Corrected the concept in the Introduction section of README.md file.
    + Added "Create a bucket" to the Feature section of README.md file.
    + Removed the code lines for checking whether a bucket with the same name exists, since AWS Java SDK works in the same way as Golang version (continuing uploading the new object to the existed bucket).
    + Added exception handling for uploadObject() and createBucket() functions. listObjects() does not throw an exception according to its implementation in AWS Java SDK.
